### PR TITLE
[FLINK-17052] [client] Introduce PlanGenerator

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*As [FLINK-16533](https://issues.apache.org/jira/browse/FLINK-16533) discussed, We move the most part logic of ExecutionEnvironment#createProgramPlan method to PlanGenerator, which can be used by ExecutionEnvironment and flink-table-planner.*

## Brief change log

  - *extract plan generation logic from ExecutionEnvironment#createProgramPlan to PlanGenerator*

## Verifying this change

This change is already covered by existing tests, such as *CollectionExecutionAccumulatorsTest, ExecutorDiscoveryAndJobClientTest*. Added PlanGeneratorTest can also verify the logic.
 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 1]
```diff
@@ -1035,6 +1030,14 @@ protected void registerCachedFilesWithPlan(Plan p) throws IOException {
 		}
 	}
 
+	/**
```
Comment (aljoscha) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: This shouldn't be exposed. You can write the test strictly without using an `ExecutionEnvironment`. Which in the end users of `PlanGenerator` probably also have to do.

Comment (godfreyhe) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: consider the following user table program:
val env = ExecutionEnvironment.getExecutionEnvironment()
val tableEnv = BatchTableEnvironment.create(env, config)
env.registerCachedFile("/tmp/cache", "my_cache")
val input = env.fromElements(...)...
tEnv.fromDataSet(input)...
tEnv.execute() // will generate plan through PlanGenerator in this method

If we do not expose this method, the generated plan does not contain the cache files registered in DataSet program.(which can be converted to Table through `BatchTableEnvironment.fromDataSet` method)

Comment (aljoscha) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: True, but I think this is tricky, because there could be any number of options that we want to forward from the `ExecutionEnvironment` but that are easy to forget. If there is a general way of forwarding everything that would probably be ok.

Comment (godfreyhe) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: I find `ExecutionConfig` has forwarded everything except `cacheFile`. CMIIW

**Comment by godfreyhe** [comment]
@flinkbot run travis

**Comment by godfreyhe** [comment]
@flinkbot run travis

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java (hunk 1 above)
Comment (godfreyhe) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: hi @aljoscha, do you have any other concern about this ?

**Comment by godfreyhe** [comment]
hi @kl0u,  do you have time to help review this pr ?

**Comment by kl0u** [comment]
Yes @godfreyhe , I will do it today.

**Comment by kl0u** [comment]
BTW both builds (Azure and Travis) seem to be failing. Have you checked them?

**Comment by godfreyhe** [comment]
that is the penultimate commit (b7adab1) failure,  Azure and Travis don't run my latest commit (bc01102). I re-trigger tests

**Comment by godfreyhe** [comment]
@flinkbot run travis

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
my private travis is passed https://travis-ci.org/github/godfreyhe/flink/builds/673799271

**Comment by kl0u** [comment]
Hehe I see.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java (hunk 1 above)
Comment (kl0u) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: From the discussion, it seems that this method was added:
1) for the tests 
2) for future usage in the `TableEnvironment`.

If this is correct, I would suggest to remove it from this PR and add it to the PR that uses it. Now it seems that this is an unused `public` method.

In addition, I agree with @aljoscha 's concerns that we may need to come up with a better and more foolproof design on how to forward stuff from the `ExecutionEnvironment` to the `TableEnvironment`. So if we achieve that, then this method would be instant legacy. 

WDYT?

Comment (godfreyhe) [reply flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java:1033]: Make sense to me, I will revert it

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+		if (config.isForceKryoEnabled() && config.isForceAvroEnabled()) {
+			LOG.warn("In the ExecutionConfig, both Avro and Kryo are enforced. Using Kryo serializer");
+		}
+		if (config.isForceKryoEnabled()) {
+			LOG.info("Using KryoSerializer for serializing POJOs");
+		}
+		if (config.isForceAvroEnabled()) {
+			LOG.info("Using AvroSerializer for serializing POJOs");
+		}
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:106]: This block here seems a bit weird, right? It may end up printing:
```
LOG.warn("In the ExecutionConfig, both Avro and Kryo are enforced. Using Kryo serializer");
LOG.info("Using KryoSerializer for serializing POJOs");
LOG.info("Using AvroSerializer for serializing POJOs");
```
which is misleading as either `kryo` or `avro` will be used.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+				public void postVisit(org.apache.flink.api.common.operators.Operator<?> visitable) {
+				}
+			});
+		}
+
+		// All types are registered now. Print information.
+		int registeredTypes = config.getRegisteredKryoTypes().size() +
+				config.getRegisteredPojoTypes().size() +
+				config.getRegisteredTypesWithKryoSerializerClasses().size() +
+				config.getRegisteredTypesWithKryoSerializers().size();
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:92]: This could go in a separate method like: `getNumberOfRegisteredTypes()`

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		int defaultKryoSerializers = config.getDefaultKryoSerializers().size() +
+				config.getDefaultKryoSerializerClasses().size();
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:94]: This could go in a separate method like: `getNumberOfDefaultKryoSerializers()`

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			LOG.debug("Registered Kryo with Serializer Classes types: {}",
+					config.getRegisteredTypesWithKryoSerializerClasses().entrySet().toString());
+			LOG.debug("Registered Kryo default Serializers: {}",
+					config.getDefaultKryoSerializers().entrySet().toString());
+			LOG.debug("Registered Kryo default Serializers Classes {}",
+					config.getDefaultKryoSerializerClasses().entrySet().toString());
+			LOG.debug("Registered POJO types: {}", config.getRegisteredPojoTypes().toString());
+
+			// print information about static code analysis
+			LOG.debug("Static code analysis mode: {}", config.getCodeAnalysisMode());
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:121]: This could go in a separate method like: `logDebuggingTypeDetails()`.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	public Plan generate() {
+		OperatorTranslation translator = new OperatorTranslation();
+		Plan plan = translator.translateToPlan(sinks, jobName);
+
+		if (config.getParallelism() > 0) {
+			plan.setDefaultParallelism(config.getParallelism());
+		}
+		plan.setExecutionConfig(config);
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:63]: This can go in a method:

```
private Plan createPlan() {
		final OperatorTranslation translator = new OperatorTranslation();
		final Plan plan = translator.translateToPlan(sinks, jobName);

		if (config.getParallelism() > 0) {
			plan.setDefaultParallelism(config.getParallelism());
		}
		plan.setExecutionConfig(config);
		return plan;
	}
```

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+	private final List<DataSink<?>> sinks;
+	private final ExecutionConfig config;
+	private final String jobName;
+
+	public PlanGenerator(
+			List<DataSink<?>> sinks,
+			ExecutionConfig config,
+			String jobName) {
+		this.sinks = sinks;
```
Comment (kl0u) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:51]: All these require `checkNotNull()` checks.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java (hunk 2 above)
Comment (godfreyhe) [reply flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java:106]: Thank for detailed check, I just copy this from `ExecutionEnvironment#createProgramPlan` method, I will correct this.

**Comment by kl0u** [comment]
@godfreyhe I just saw you last commit, I think it was a nice change to pass the `cacheFile` in the constructor of the `PlanGenerator` and register the files there. This delegates the whole process of the "plan creation" to the new class.

My comment was just to not expose it through the new public method :)

**Comment by godfreyhe** [comment]
> @godfreyhe I just saw you last commit, I think it was a nice change to pass the `cacheFile` in the constructor of the `PlanGenerator` and register the files there. This delegates the whole process of the "plan creation" to the new class.
> 
> My comment was just to not expose it through the new public method :)

oh, I misunderstand you concern. `PlanGenerator` also supports  cache registration. but for `TableEnvironment`, the argument of cacheFiles are alway empty, even if `ExecutionEnvironment` has cache files.

File: flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java [hunk 8]
```diff
@@ -91,8 +91,6 @@ private TestUtils() {}
 	 * @param expectedParallelism expected operator parallelism
 	 */
 	public static void verifyParallelism(ExecutionEnvironment env, int expectedParallelism) {
-		env.setParallelism(2 * expectedParallelism);
```
Comment (kl0u) [inline flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: Is this change needed?

Comment (godfreyhe) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: yes. before this pr, `CollectionEnvironment` override `getParallelism` method, and always return 1.  `ExecutionEnvironment#createProgramPlan` get parallelism through `ExecutionEnvironment#getParallelism`. Now, `PlanGenerator` get parallelism through `config.getParallelism()`.  When `CollectionEnvironment` is created in `ExecutionEnvironment#createCollectionsEnvironment` method, the parallelism in config will be set as 1, while after `env.setParallelism(2 * expectedParallelism);` is called, the parallelism in config was changed to 4. which is unexpected.

Comment (kl0u) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: I agree that the test design is a bit confusing. But doesn't this mean that this PR changes existing behavior?

Comment (godfreyhe) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: I think so. `CollectionEnvironment` is a special case that requires the parallelism is always 1.

**Comment by godfreyhe** [comment]
My private travis is green  https://travis-ci.org/github/godfreyhe/flink/builds/674785749

File: flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java (hunk 8 above)
Comment (godfreyhe) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: hi @kl0u does this make sense to you ?

Comment (kl0u) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: So given that users expect a specific behaviour, I think we should not change it, as this may break user code. WDYT?

Comment (godfreyhe) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: One approach is we add a constructor parameter named `defaultParallelism` for `PlanGenerator`, give the value from `ExecutionEnvironment#getParallelism`. That does not  brake current behavior, but that also means the parallelism from `ExecutionEnvironment` and from `ExecutionConfig` may be different. WDYT ?

Comment (godfreyhe) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: hi @kl0u I have reverted this change. do you have other concern ?

Comment (kl0u) [reply flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/generator/TestUtils.java:94]: @godfreyhe I agree that the propagation of the parallelism is confusing. But I think that we should not change existing behaviour in order to avoid users having unexpected results. I am having a look now at your PR.

**Comment by godfreyhe** [comment]
> Changes look good @godfreyhe ! I will merge as soon as my Travis gives green.

Thanks you for your patient suggestion!

**Comment by kl0u** [comment]
Thanks for the contribution @godfreyhe !
